File: src/server/seeding/demo-board.ts
import { db } from "~/server/db";
import { boards, columns, cards } from "~/server/db/schema";
import { generateDemoBoardContent } from "./demo-content";

/**
 * Create the demo board with all its content for the Kanban Admin
 */
export async function createDemoBoard(adminUserId: string): Promise<string> {
  console.log("🎨 Creating demo board with engaging content...");

  try {
    const demoContent = generateDemoBoardContent();

    // Create the demo board
    const [demoBoard] = await db
      .insert(boards)
      .values({
        title: demoContent.boardTitle,
        description: demoContent.boardDescription,
        userId: adminUserId,
        isArchived: false,
      })
      .returning();

    if (!demoBoard) {
      throw new Error("Failed to create demo board");
    }

    console.log(`📋 Demo board created: "${demoBoard.title}"`);

    // Create columns and cards in a transaction for consistency
    await db.transaction(async (tx) => {
      for (const columnData of demoContent.columns) {
        // Create the column
        const [column] = await tx
          .insert(columns)
          .values({
            boardId: demoBoard.id,
            name: columnData.name,
            position: columnData.position,
            isCollapsed: false,
          })
          .returning();

        if (!column) {
          throw new Error(`Failed to create column: ${columnData.name}`);
        }

        console.log(`  📝 Created column: "${column.name}" with ${columnData.cards.length} cards`);

        // Create cards for this column
        if (columnData.cards.length > 0) {
          await tx.insert(cards).values(
            columnData.cards.map((cardData) => ({
              columnId: column.id,
              title: cardData.title,
              description: cardData.description,
              position: cardData.position,
              createdBy: adminUserId,
            }))
          );
        }
      }
    });

    console.log("✅ Demo board creation completed successfully!");
    return demoBoard.id;
  } catch (error) {
    console.error("❌ Failed to create demo board:", error);
    throw error;
  }
}

File: api/app/seeding/demo_board.py
from sqlalchemy.orm import Session

from app.models import Board, BoardColumn, Card
from app.seeding.demo_content import generate_demo_board_content


def create_demo_board(db: Session, board_owner_id: str, card_creator_id: str) -> str:
    """Create the demo board with all its columns and cards. Returns board id."""
    content = generate_demo_board_content()

    board = Board(
        title=content["boardTitle"],
        description=content["boardDescription"],
        user_id=board_owner_id,
        is_archived=False,
    )
    db.add(board)
    db.flush()

    for column_data in content["columns"]:
        column = BoardColumn(
            board_id=board.id,
            name=column_data["name"],
            position=column_data["position"],
            is_collapsed=False,
        )
        db.add(column)
        db.flush()

        for card_data in column_data["cards"]:
            db.add(
                Card(
                    column_id=column.id,
                    title=card_data["title"],
                    description=card_data["description"],
                    position=card_data["position"],
                    created_by=card_creator_id,
                )
            )

    db.commit()
    return board.id
